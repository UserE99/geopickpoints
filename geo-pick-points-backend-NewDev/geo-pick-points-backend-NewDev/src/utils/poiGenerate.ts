import * as GameService from "../services/GameService";
import { GameResource, POIListResource, POIResource } from "src/Resources";
import * as POIService from "../services/POIService";
import * as POILISTSService from "../services/POIListsService"
import { Types } from "mongoose";

export async function createExampleGame() {
    try {
        // Definiere alle POIs
        const berlinPOIs: POIResource[] = [
            { name: "Alexanderplatz", lat: 52.520008, long: 13.404954, beschreibung: "Ein belebter Platz mit Fernsehturm, Geschäften und urbanem Flair.", punkte: 100 },
            { name: "Brandenburger Tor", lat: 52.516275, long: 13.377704, beschreibung: "Ein ikonisches Monument und Symbol für Geschichte und Einheit.", punkte: 200 },
            { name: "Podsdamer Platz", lat: 52.509290, long: 13.376340, beschreibung: "Ein moderner Knotenpunkt mit Architektur, Kultur und Unterhaltung.", punkte: 100 },
            { name: "Oberbaumbrücke", lat: 52.501834, long: 13.445656, beschreibung: "Eine markante Brücke mit Doppeldeck-Architektur und historischem Charme.", punkte: 100 },
            { name: "Museumsinsel", lat: 52.516260, long: 13.402480, beschreibung: "Ein einzigartiges Kulturensemble mit weltberühmten Museen.", punkte: 100 },
            { name: "Volkspark Friedrichhain", lat: 52.528730, long: 13.442284, beschreibung: "Ein weitläufiger Park mit grünen Wiesen, Hügeln und Entspannungsoasen.", punkte: 50 },
            { name: "Deutsches Technikmuseum", lat: 52.498603, long: 13.378154, beschreibung: "Ein faszinierendes Museum mit historischen Exponaten zu Technik und Ingenieurskunst.", punkte: 50 },
            { name: "Checkpoint Charlie", lat: 52.507530, long: 13.390378, beschreibung: "Ein historischer Grenzpunkt und Symbol des Kalten Krieges.", punkte: 200 },
        ];

        const bhtPOIs: POIResource[] = [
            { name: "Workout Park", lat: 52.545374, long: 13.352802, beschreibung: "Workout Park", punkte: 50 },
            { name: "Spielplatz auf dem Zeppelinplatz", lat: 52.546413, long: 13.353094, beschreibung: "Spielplatz auf dem Zeppelinplatz", punkte: 200 },
            { name: "Einfahrt", lat: 52.546101, long: 13.355068, beschreibung: "Einfahrt", punkte: 50 },
            { name: "Fahrradständer Zeppelinplatz", lat: 52.545717, long: 13.351990, beschreibung: "Fahrradständer Zeppelinplatz", punkte: 100 },
            { name: "Eingang Zeppelinplatz", lat: 52.545879, long: 13.354316, beschreibung: "Eingang Zeppelinplatz", punkte: 100 },
        ];

        const potsdamPOIs: POIResource[] = [
            { name: "Park Sanssouci", lat: 52.40340322194994, long: 13.029932869765789, beschreibung: "Park Sanssouci", punkte: 50 },
            { name: "Filmpark Babelsberg", lat: 52.38494249076188, long: 13.11788422883501, beschreibung: "Filmpark Babelsberg", punkte: 200 },
            { name: "Glienicker Brücke", lat: 52.413617910656896, long: 13.090731035582499, beschreibung: "Glienicker Brücke", punkte: 50 },
            { name: "Schloss Cecilienhof", lat: 52.41958918275413, long: 13.070846179757533, beschreibung: "Schloss Cecilienhof", punkte: 100 },
            { name: "Biosphäre Potsdam", lat: 52.41901031179306, long: 13.049338804902431, beschreibung: "Biosphäre Potsdam", punkte: 100 },
        ];

        // Erstelle alle POIs und sammle die IDs
        const createPOIs = async (pois: POIResource[]): Promise<string[]> => {
            const poiIds: string[] = [];
            for (const poi of pois) {
                const fullData = await POIService.createPOI(poi);
                poiIds.push(fullData.id!);
            }
            return poiIds;
        };

        const berlinPoiIds = await createPOIs(berlinPOIs);
        const bhtPoiIds = await createPOIs(bhtPOIs);
        const potsdamPoiIds = await createPOIs(potsdamPOIs);

        // Spiel erstellen
        const gameData: GameResource = {
            title: "Berlin Sehenswürdigkeiten",
            beschreibung: "Einige der bekanntesten Sehenswürdigkeiten in Berlin",
            poilId: [...berlinPoiIds, ...bhtPoiIds],
            maxTeam: 5,
            userId: new Types.ObjectId().toString(),
        };
        await GameService.createGame(gameData);

        const gameDataPotsdam: GameResource = {
            title: "Potsdam Sehenswürdigkeiten",
            beschreibung: "Einige der bekanntesten Sehenswürdigkeiten in Potsdam",
            poilId: [...potsdamPoiIds],
            maxTeam: 5,
            userId: new Types.ObjectId().toString(),
        };
        await GameService.createGame(gameDataPotsdam);


        //1.  POI-Liste Berlin erstellen
        const poiList: POIListResource = {
            name: "Berlin",
            poilId: [...berlinPoiIds, ...bhtPoiIds],
        };
        await POILISTSService.createPOIList(poiList);

        //2.  POI-Liste Potsdam erstellen
        const poiListPotsdam: POIListResource = {
            name: "Potsdam",
            poilId: [...potsdamPoiIds],
        };
        await POILISTSService.createPOIList(poiListPotsdam);

        console.log("Beispielspiel und POI-Liste erfolgreich erstellt.");
    } catch (error: any) {
        console.error("Fehler beim Erstellen des Beispielspiels:", error.message);
        throw error;
    }
}
