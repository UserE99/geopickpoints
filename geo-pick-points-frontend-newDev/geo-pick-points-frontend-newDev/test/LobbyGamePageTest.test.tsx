import React, { act } from 'react';
import { render, screen, fireEvent, waitFor, within } from '@testing-library/react';
import { Provider } from 'react-redux';
import "@testing-library/jest-dom";
import { BrowserRouter, MemoryRouter, useLocation } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import LobbyGamePage from '../src/Pages/LobbyGamePage';
import * as WebSocketSetup from '../src/utils/WebSocketSetup';
import fetchMock from "jest-fetch-mock";
import { thunk } from 'redux-thunk';
import { joinInTeam } from '../src/actions/JoinTeamGame';

fetchMock.enableMocks();


jest.mock('../src/utils/WebSocketSetup', () => ({
    __esModule: true,
    default: jest.fn(() => ({
        messages: [],
        sendMessage: jest.fn(),
    })),
}));

const mockStore = configureStore([]);
const mockedLocationState = {
    state: {
        nickName: 'TestPlayer',
        playerID: '1234',
        codeInvite: 'testCode',
        teamID: null,
    },
};

jest.mock("../src/actions/JoinTeamGame", () => {
    const actualModule = jest.requireActual("../src/actions/JoinTeamGame");

    return {
        ...actualModule,
        joinInTeam: Object.assign(jest.fn(), actualModule.joinInTeam),
    };
});

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn(),
}));

describe('LobbyGamePage', () => {
    let store: ReturnType<typeof mockStore>;

    beforeEach(() => {
        store = mockStore({
            gameStatus: { isGameStarted: false },
        });
        (useLocation as jest.Mock).mockReturnValue(mockedLocationState);
    });

    it('renders the LobbyGamePage component and matches snapshot', async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <LobbyGamePage />
                </MemoryRouter>
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('displays the loading state initially', async () => {
        jest.spyOn(global, 'fetch').mockResolvedValueOnce({
            ok: true,
            json: async () => [],
        } as Response);

        store = mockStore({
            teams: null,
            gameStatus: { isGameStarted: false },
        });

        render(
            <Provider store={store}>
                <MemoryRouter>
                    <LobbyGamePage />
                </MemoryRouter>
            </Provider>
        );


        expect(screen.getByText((_, element) => {
            const hasText = (text: string) => text.includes("Loading teams...");
            const childrenDontHaveText = Array.from(element?.children || []).every(
                (child) => !hasText(child.textContent || "")
            );
            return hasText(element?.textContent || "") && childrenDontHaveText;
        })).toBeInTheDocument();
    });


    it('displays an error message if fetching team data fails', async () => {
        jest.spyOn(global, 'fetch').mockRejectedValueOnce(new Error('Failed to fetch team data'));

        await act(async () => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <LobbyGamePage />
                    </MemoryRouter>
                </Provider>
            );
        });

        await waitFor(() => {
            expect(screen.getByText(/Could not load teams. Please try again./i)).toBeInTheDocument();
        });

        global.fetch.mockRestore();
    });


    it('handles joining and leaving a team', async () => {
        const store = mockStore({
            gameStatus: { isGameStarted: false },
            playerTeamInfo: { teamId: null, nickName: null },
        });

        // Dispatch als Mock-Funktion setzen
        store.dispatch = jest.fn();
        const mockSendMessage = jest.fn();
        (WebSocketSetup.default as jest.Mock).mockReturnValue({
            messages: [],
            sendMessage: mockSendMessage,
        });

        const mockTeamData = [
            {
                _id: 'team1',
                name: 'Team One',
                shortName: 'T1',
                players: [
                    { nickName: 'TestPlayer', playerId: '1234' },
                ],
            },
        ];

        jest.spyOn(global, 'fetch').mockResolvedValueOnce({
            ok: true,
            json: async () => mockTeamData,
        } as Response);

        render(
            <Provider store={store}>
                <MemoryRouter>
                    <LobbyGamePage />
                </MemoryRouter>
            </Provider>
        );

        // Wait for the teams to load
        await waitFor(() => expect(screen.getByText(/Select your Teams/i)).toBeInTheDocument());
        // Verify the Host Name
        const hostNameContainer = screen.getByText(/Host Name :/i).closest('h5');
        expect(
            within(hostNameContainer ?? document.createElement('div')).getByText(/TestPlayer/i)
        ).toBeInTheDocument();
        // "Join Team" Button klicken
        const joinButton = await screen.findByRole('button', { name: /Join Team/i });
        fireEvent.click(joinButton);
        console.log("Button clicked!");

        // Prüfen, ob `dispatch` nach dem Klick aufgerufen wurde
        console.log("Dispatched Actions nach Click:", store.getActions());

        // Überprüfe, ob `store.dispatch` aufgerufen wurde
        await waitFor(() => {
            expect(store.dispatch).toHaveBeenCalledWith(expect.objectContaining({
                type: joinInTeam.pending.type,  //  hier wird joinInTeam.pending.type verwendet
                payload: { teamId: 'team1', nickName: 'TestPlayer' },
            }));
        });


        // Hol  alle Dispatch-Aktionen
        const actions = store.getActions();
        expect(actions).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    type: joinInTeam.fulfilled.type, // Redux Toolkit Action-Type
                    payload: { teamId: 'team1', nickName: 'TestPlayer' },
                }),
            ])
        );

        // Stelle sicher, dass "Leave Team" Button sichtbar wird
        await waitFor(() => {
            expect(screen.getByRole('button', { name: /Leave Team/i })).toBeInTheDocument();
        });

        // Simuliere Klick auf "Leave Team"
        const leaveButton = screen.getByRole('button', { name: /Leave Team/i });
        fireEvent.click(leaveButton);

        // Stelle sicher, dass die korrekten Redux-Aktionen gesendet wurden
        await waitFor(() => {
            expect(store.dispatch).toHaveBeenCalledWith(expect.objectContaining({
                type: 'LEAVE_TEAM', // Falls eine separate "leave" Action hast
                payload: { teamId: 'team1', nickName: 'TestPlayer' },
            }));
        });

        // Prüfe, ob `mockSendMessage` aufgerufen wurde (WebSocket Nachricht)
        expect(mockSendMessage).toHaveBeenCalledWith({
            type: 'leave',
            playerId: '1234',
            playerName: 'TestPlayer',
            teamId: 'team1',
        });

        // API-Mocks zurücksetzen
        global.fetch.mockRestore();
    });

});
