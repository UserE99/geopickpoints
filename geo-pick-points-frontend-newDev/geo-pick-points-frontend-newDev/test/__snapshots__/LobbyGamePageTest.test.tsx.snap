// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbyGamePage renders the LobbyGamePage component and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mt-5"
  >
    <h2>
      Lobby Host : TestPlayer , you are here in the lobby game.
    </h2>
    <p>
      Loading teams...
    </p>
  </div>
</DocumentFragment>
`;
